src/components/implementation/fork_tests/voter/voter.c

```diff
@@ -0,0 +1,290 @@
+			return -1; // whatever, shouldn't happen if blocked
+		}
+		else {
+			c->have_B2A_data = 1;
+			c->B2A_data = data;
+			ret = block_replica(replica);
+			if (ret < 0) printc("Error\n");
+			/* On wakeup */	
+			return 0;
+		}
+	}
+	else {
+		BUG();
+		return -1; // shut up compiler
+	}
+}
+
+int nread(spdid_t spdid, replica_type from, int data) {
+	struct replica_info *replica = get_replica(spdid);
+	struct nmod_comp *component = get_component(spdid);
+	struct nmod_comp *from_comp = (from == ping) ? &components[0] : (from == pong) ? & components[1] : NULL;
+	struct channel *c = get_channel(component, from_comp);
+	int ret;
+	if (!replica) BUG();
+	if (!c) BUG();
+
+	if (component == c->A) {
+		if (c->have_B2A_data) {
+			c->have_B2A_data = 0;
+			return c->B2A_data;
+		}
+		else {
+			ret = block_replica(replica);
+			if (ret < 0) printc("Error");
+			/* On wakeup */	
+			assert(c->have_B2A_data);
+			c->have_B2A_data = 0; // well this is wrong because maybe other replicas still need to read
+			return c->B2A_data;
+		}
+	}
+	else if (component == c->B) {
+		if (c->have_A2B_data) {
+			c->have_A2B_data = 0;
+			return c->A2B_data;
+		}
+		else {
+			ret = block_replica(replica);
+			if (ret < 0) printc("Error");
+			/* On wakeup */	
+			assert(c->have_A2B_data);
+			c->have_A2B_data = 0;
+			return c->A2B_data;
+		}
+	}
+	else {
+		BUG();
+		return -1; // shut up compiler
+	}
+}
+
+int confirm(spdid_t spdid, replica_type type) {
+	printc("Confirming readiness of spdid %d\n", spdid);
+	
+	int ret;
+	int i;
+	
+	int t = (type == ping) ? 0 : 1;	
+	struct replica_info *replicas = components[t].replicas;
+	
+	for (i = 0; i < components[t].nreplicas; i++) {
+		if (replicas[i].spdid == 0) {
+			printc("creating replica for spdid %d in slot %d\n", spdid, i);
+			replicas[i].spdid = spdid;
+			//replicas[i].spdid = quarantine_fork(cos_spd_id(), comp2fork);
+			//if (replicas[i].spdid == 0) printc("Error: f1 fork failed\n");
+			replicas[i].buf_read = cbuf_alloc(1024, &replicas[i].read_buffer); 
+			replicas[i].buf_write = cbuf_alloc(1024, &replicas[i].write_buffer);
+			replicas[i].blocked = 0;
+			replicas[i].thread_id = cos_get_thd_id();
+
+			map[spdid] = type;
+			return 0;
+		}
+		else {
+			printc("replica for spdid %d already exists\n", replicas[i].spdid);
+		}
+	}
+	return 0;
+}
+
+void monitor(void) {
+	struct nmod_comp *rcv_comp;
+	struct replica_info *rcv_replica;
+	int i, j;
+	int found = 0;
+
+	for (i = 0; i < n_channels; i++) {
+		printc("Examing channel %d between nmodcomps with replicas %d and %d \n", i, channels[i].A->replicas[0].spdid, channels[i].B->replicas[0].spdid);
+
+		printc("channel A2B_data present %d (%d)\n", channels[i].have_A2B_data, channels[i].A2B_data);
+		printc("channel B2A_data present %d (%d)\n", channels[i].have_B2A_data, channels[i].B2A_data);
+
+		if (channels[i].have_A2B_data) {
+			rcv_comp = channels[i].B;
+			found = 1;
+		}
+		else if (channels[i].have_B2A_data) {
```

Comment:
Wait, how are you also viewing the channel commits? I didn't include those in the PR??? Must be a github feature. But this is fantastic, you basically seem to have done this PR and the next one at once.